## Hard Drive Failure Analytics Analytics - Exploratory Data Analysis
Purpose: The purpose of this notebook is to review the source data and create step by step process and extracting the data from online source and reviewing the data locally

Objectives:
- Extract zipfiles from source
- Uzip files and import CSV files
- Describe data (columns, schema, datatypes, total row count, column count etc.)

In [2]:
from glob import glob
import os
import pandas as pd
import pyarrow.parquet as pq
import requests
from sqlalchemy import create_engine
import zipfile

Backblaze datasets are organized as the following:
- 2013/2014 - full annual zip file with multiple csv files. Url suffix ends as **data_{year}.zip**
- 2015 and onward - separated as quarterly zip files with multiple csv files (4 quarters for each year). Url suffix ends as **data_{quarter}_{year}.zip**

For testing let's pull just 2025 and see what we have.

Minigoals:
- Extract one zip file at time -> reduce number of data fields -> read csv files as one

In [3]:
year = 2025
qtr = "1"
file_suffix = f"{"data_Q"+qtr}_{year}"

In [4]:
# Creating helpful functions

def download_and_extract(zip_url, extract_dir='../data/zip'):    
    # 1. Ensure ./data/zip folder exists
    os.makedirs(extract_dir, exist_ok=True)

    # Get filename from URL path
    zip_filename = zip_url.split('/')[-1] or 'archive.zip'

    # 2. Download zip file from URL and save to extract_dir
    zip_path = os.path.join(extract_dir, zip_filename)
    
    print(f"Downloading {zip_url} to {zip_path}...")
    response = requests.get(zip_url, stream=True)
    response.raise_for_status()  # Raise HTTPError for bad responses
    
    with open(zip_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            if chunk:
                f.write(chunk)
    
    return zip_path

def unzip_file(input_file, output_dir=f"../data/raw/{year}"):
    # Ensure output directory exists
    os.makedirs(output_dir, exist_ok=True)
    
    # Validate input file exists
    if not os.path.isfile(input_file):
        raise FileNotFoundError(f"Input file does not exist: {input_file}")
    
    with zipfile.ZipFile(input_file, 'r') as zip_ref:
        # List contents (optional check)
        print("Contents:", zip_ref.namelist())
        
        # Extract all files to the specified output directory
        zip_ref.extractall(output_dir)
        print(f"Extracted to: {output_dir}")


def concatenate_files_to_df(folder_path: str, file_pattern: str = "*.csv") -> pd.DataFrame:
    """
    Efficiently concatenates all CSV files from a folder into one DataFrame.
    
    Args:
        folder_path (str): Path to directory containing CSV files
        file_pattern (str): Glob pattern for matching files (default: *.csv)
    
    Returns:
        pd.DataFrame: Concatenated DataFrame
    """
    # 1. Collect all matching CSV file paths first (avoid repeated glob calls)
    csv_files = sorted(glob(os.path.join(folder_path, file_pattern)))
    
    if not csv_files:
        raise FileNotFoundError(f"No CSV files found in {folder_path}")
    
    # 2. Concatenate with memory-efficient parameters
    df_concatenated = pd.concat(
        [pd.read_csv(file) for file in csv_files],
        ignore_index=True,          # Resets index so they start at 0
        sort=False                  # Skips sorting (faster, assume consistent schema)
    )
    
    return df_concatenated


In [ ]:
base_url = "https://f001.backblazeb2.com/file/Backblaze-Hard-Drive-Data/"

In [6]:
# Extract data
# download_and_extract(base_url+f"{file_suffix}.zip")

In [7]:
# unzip_file(f"../data/zip/{file_suffix}.zip")

In [8]:
# Read quick sample
df_test = pd.read_csv(f"../data/raw/{year}/{file_suffix}/{year}-01-01.csv"
    , nrows=100)
df_test.head()

,date,serial_number,model,capacity_bytes,failure,datacenter,cluster_id,vault_id,pod_id,pod_slot_num,...,smart_250_normalized,smart_250_raw,smart_251_normalized,smart_251_raw,smart_252_normalized,smart_252_raw,smart_254_normalized,smart_254_raw,smart_255_normalized,smart_255_raw
0,2025-01-01,2207E60CC65A,CT250MX500SSD1,250059350016,0,sac0,0,1028,13,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2025-01-01,2340E87B92B5,CT250MX500SSD1,250059350016,0,sac0,0,1028,14,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2025-01-01,2EGK64VX,HGST HUH728080ALE604,8001563222016,0,sac0,0,1028,4,12.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2025-01-01,2EHZAKAX,HGST HUH728080ALE604,8001563222016,0,sac0,0,1028,12,30.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2025-01-01,2EJ02A1X,HGST HUH728080ALE604,8001563222016,0,sac0,0,1028,10,14.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [9]:
df_test.dtypes

date                        str
serial_number               str
model                       str
capacity_bytes            int64
failure                   int64
                         ...   
smart_252_raw           float64
smart_254_normalized    float64
smart_254_raw           float64
smart_255_normalized    float64
smart_255_raw           float64
Length: 197, dtype: object

In [10]:
df_test.shape

(100, 197)

## Notes
- The csv file sizes are too large to combine as one df for an entire quarter of data. Load into postgres in chunks for analysis instead.
- Since we are focused on failure rates, there are specific SMART attributes that were identifed as key to indicating failure or soon to be failure drives (refernced from backblaze):
    - SMART 5: Reallocated_Sector_Count.
	- SMART 187: Reported_Uncorrectable_Errors.
	- SMART 188: Command_Timeout.
	- SMART 197: Current_Pending_Sector_Count.
	- SMART 198: Offline_Uncorrectable.

In [ ]:
dtype = {
    "serial_number": "string",
    "model": "string",
    "capacity_bytes": "Int64",
    "failure": "Int64",
    "datacenter": "string",
    "cluster_id": "Int64",
    "vault_id": "Int64",
    "pod_id": "Int64",
    "pod_slot_num": "float64",
    "is_legacy_format": "boolean",
    "smart_5_normalized": "float64",  # SMART attribute (normalized value)
    "smart_5_raw": "float64", # Raw SMART value
    "smart_187_normalized": "float64",
    "smart_187_raw": "float64",
    "smart_188_normalized": "float64",
    "smart_188_raw": "float64",
    "smart_197_normalized": "float64",
    "smart_197_raw": "float64",
    "smart_198_normalized": "float64",
    "smart_198_raw": "float64"
}

parse_dates = [
    "date"
]

In [12]:
col_to_keep = []
for col in df_test.columns:
    if col in dtype or col in parse_dates:
        col_to_keep.append(col)
print(col_to_keep)

['date', 'serial_number', 'model', 'capacity_bytes', 'failure', 'datacenter', 'cluster_id', 'vault_id', 'pod_id', 'pod_slot_num', 'is_legacy_format', 'smart_5_normalized', 'smart_5_raw', 'smart_187_normalized', 'smart_187_raw', 'smart_188_normalized', 'smart_188_raw', 'smart_197_normalized', 'smart_197_raw', 'smart_198_normalized', 'smart_198_raw']


In [13]:
df = pd.read_csv(
    f"../data/raw/{year}/{file_suffix}/{year}-01-01.csv"
    , nrows=100
    , dtype=dtype
    ,parse_dates=parse_dates
    ,usecols=col_to_keep
)
df.head(10)

,date,serial_number,model,capacity_bytes,failure,datacenter,cluster_id,vault_id,pod_id,pod_slot_num,...,smart_5_normalized,smart_5_raw,smart_187_normalized,smart_187_raw,smart_188_normalized,smart_188_raw,smart_197_normalized,smart_197_raw,smart_198_normalized,smart_198_raw
0,2025-01-01,2207E60CC65A,CT250MX500SSD1,250059350016,0,sac0,0,1028,13,NaN,...,100.0,0.0,100.0,0.0,NaN,NaN,100.0,0.0,100.0,0.0
1,2025-01-01,2340E87B92B5,CT250MX500SSD1,250059350016,0,sac0,0,1028,14,NaN,...,100.0,0.0,100.0,0.0,NaN,NaN,100.0,0.0,100.0,0.0
2,2025-01-01,2EGK64VX,HGST HUH728080ALE604,8001563222016,0,sac0,0,1028,4,12.0,...,100.0,0.0,NaN,NaN,NaN,NaN,100.0,0.0,100.0,0.0
3,2025-01-01,2EHZAKAX,HGST HUH728080ALE604,8001563222016,0,sac0,0,1028,12,30.0,...,100.0,0.0,NaN,NaN,NaN,NaN,100.0,0.0,100.0,0.0
4,2025-01-01,2EJ02A1X,HGST HUH728080ALE604,8001563222016,0,sac0,0,1028,10,14.0,...,100.0,0.0,NaN,NaN,NaN,NaN,100.0,0.0,100.0,0.0
5,2025-01-01,7LZ021LA,Seagate BarraCuda SSD ZA250CM10002,250059350016,0,sac0,0,1028,6,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,2025-01-01,S2ZYJ9CF511681,ST500LM012 HN,500107862016,0,sac0,0,1028,10,NaN,...,252.0,0.0,NaN,NaN,NaN,NaN,252.0,0.0,252.0,0.0
7,2025-01-01,S2ZYJ9GGB01000,ST500LM012 HN,500107862016,0,sac0,0,1028,0,NaN,...,252.0,0.0,NaN,NaN,NaN,NaN,252.0,0.0,252.0,0.0
8,2025-01-01,S2ZYJ9GGB01001,ST500LM012 HN,500107862016,0,sac0,0,1028,4,NaN,...,252.0,0.0,NaN,NaN,NaN,NaN,252.0,0.0,252.0,0.0
9,2025-01-01,S2ZYJ9GGB01020,ST500LM012 HN,500107862016,0,sac0,0,1028,2,NaN,...,252.0,0.0,NaN,NaN,NaN,NaN,252.0,0.0,252.0,0.0


In [14]:
df.dtypes

date                    datetime64[us]
serial_number                   string
model                           string
capacity_bytes                   Int64
failure                          Int64
datacenter                      string
cluster_id                       Int64
vault_id                         Int64
pod_id                           Int64
pod_slot_num                   float64
is_legacy_format               boolean
smart_5_normalized             float64
smart_5_raw                    float64
smart_187_normalized           float64
smart_187_raw                  float64
smart_188_normalized           float64
smart_188_raw                  float64
smart_197_normalized           float64
smart_197_raw                  float64
smart_198_normalized           float64
smart_198_raw                  float64
dtype: object

### Setting up postgres database

In [15]:
# Create sql engine
engine = create_engine('postgresql+psycopg://root:root@localhost:5432/hard_drive_db')

In [16]:
print(pd.io.sql.get_schema(df, name='hard_drive_data', con=engine))


CREATE TABLE hard_drive_data (
	date TIMESTAMP WITHOUT TIME ZONE, 
	serial_number TEXT, 
	model TEXT, 
	capacity_bytes BIGINT, 
	failure BIGINT, 
	datacenter TEXT, 
	cluster_id BIGINT, 
	vault_id BIGINT, 
	pod_id BIGINT, 
	pod_slot_num FLOAT(53), 
	is_legacy_format BOOLEAN, 
	smart_5_normalized FLOAT(53), 
	smart_5_raw FLOAT(53), 
	smart_187_normalized FLOAT(53), 
	smart_187_raw FLOAT(53), 
	smart_188_normalized FLOAT(53), 
	smart_188_raw FLOAT(53), 
	smart_197_normalized FLOAT(53), 
	smart_197_raw FLOAT(53), 
	smart_198_normalized FLOAT(53), 
	smart_198_raw FLOAT(53)
)




In [17]:
# Create empty table with data fields
first = True
if first:
    df.head(n=0).to_sql(name='hard_drive_data', con=engine, if_exists='replace')
    print("Table created")
    first = False

0

In [32]:
files = sorted(os.listdir(f"../data/raw/{year}/{file_suffix}/"))
print(files[:3])

['2025-01-01.csv', '2025-01-02.csv', '2025-01-03.csv']


In [ ]:
# Ingest data in file chunks
for data_file in files:
    df = pd.read_csv(
        f"../data/raw/{year}/{file_suffix}/{data_file}",
        dtype=dtype,
        parse_dates=parse_dates,
        usecols = col_to_keep
    )
    df.to_sql(name='hard_drive_data', con=engine, if_exists='append')
    print(f"Loaded {len(df)} records from file: {data_file}")

Loaded 304957 records from file: 2025-01-01.csv
Loaded 304842 records from file: 2025-01-02.csv
Loaded 304848 records from file: 2025-01-03.csv
Loaded 304922 records from file: 2025-01-04.csv
Loaded 304897 records from file: 2025-01-05.csv
Loaded 305135 records from file: 2025-01-06.csv
Loaded 305211 records from file: 2025-01-07.csv
Loaded 305243 records from file: 2025-01-08.csv
Loaded 305248 records from file: 2025-01-09.csv
Loaded 305255 records from file: 2025-01-10.csv
Loaded 305156 records from file: 2025-01-11.csv
Loaded 305017 records from file: 2025-01-12.csv
Loaded 305072 records from file: 2025-01-13.csv
Loaded 305230 records from file: 2025-01-14.csv
Loaded 305105 records from file: 2025-01-15.csv
Loaded 305074 records from file: 2025-01-16.csv
Loaded 305186 records from file: 2025-01-17.csv
Loaded 305141 records from file: 2025-01-18.csv
Loaded 305105 records from file: 2025-01-19.csv
Loaded 305122 records from file: 2025-01-20.csv
Loaded 305113 records from file: 2025-01

### Creating One Ingestion Script

In [47]:
# refactor code to accept year and quarter parameters
def download_data(year=2025, qtr=1):    
    # Ensure ../data/zip folder exists
    download_dir='../data/zip'
    os.makedirs(download_dir, exist_ok=True)

    # Define base url
    base_url = "https://f001.backblazeb2.com/file/Backblaze-Hard-Drive-Data"

    # File structure
    file_suffix = "data_Q"+f"{qtr}_{year}"

    zip_url = f"{base_url}/{file_suffix}.zip"
    # Get filename from URL path
    zip_filename = zip_url.split('/')[-1] or 'archive.zip'

    # Download zip file from URL and save to download_dir
    zip_path = os.path.join(download_dir, zip_filename)
    
    print(f"Downloading {zip_url} to {zip_path}...")
    response = requests.get(zip_url, stream=True)
    response.raise_for_status()  # Raise HTTPError for bad responses
    
    with open(zip_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            if chunk:
                f.write(chunk)
    print(f"{zip_path} is downloaded!")
    return zip_path

In [48]:
download_data_path = download_data(year=2025, qtr=1)
print(download_data_path)

../data/zip/data_Q1_2025.zip is downloaded!
../data/zip/data_Q1_2025.zip


In [82]:
# Data must be structured as data_{qtr}_{year}.zip
print(download_data_path.split('/')[-1].split('.')[0])

data_Q1_2025


In [ ]:
# refactor code to accept previously generated download file path
def unzip_file(input_file=download_data_path):
    
    # Extract year and qtr from zip_path to create folder names
    # Note: file name must be structured as ../data/zip/data_{qtr}_{year}.zip for year/qtr to work
    zip_path_year = download_data_path.split('/')[-1].split('_')[-1].split('.')[0]
    # zip_path_qtr = download_data_path.split('/')[-1].split('_')[1]

    # Ensure extract directory exists
    extract_dir=f"../data/source_csv/{zip_path_year}"
    os.makedirs(extract_dir, exist_ok=True)
    
    # Validate input file exists
    if not os.path.isfile(input_file):
        raise FileNotFoundError(f"Input file does not exist: {input_file}")
    
    with zipfile.ZipFile(input_file, 'r') as zip_ref:
        # List contents (optional check)
        print("Contents:", zip_ref.namelist())
        
        # Extract all files to the specified extract directory
        zip_ref.extractall(extract_dir)
        print(f"A total of {len(zip_ref.namelist())} file(s) extracted.") 
        print(f"Extracted to: {extract_dir}")

In [88]:
unzip_file()

Contents: ['data_Q1_2025/2025-01-01.csv', 'data_Q1_2025/2025-01-02.csv', 'data_Q1_2025/2025-01-03.csv', 'data_Q1_2025/2025-01-04.csv', 'data_Q1_2025/2025-01-05.csv', 'data_Q1_2025/2025-01-06.csv', 'data_Q1_2025/2025-01-07.csv', 'data_Q1_2025/2025-01-08.csv', 'data_Q1_2025/2025-01-09.csv', 'data_Q1_2025/2025-01-10.csv', 'data_Q1_2025/2025-01-11.csv', 'data_Q1_2025/2025-01-12.csv', 'data_Q1_2025/2025-01-13.csv', 'data_Q1_2025/2025-01-14.csv', 'data_Q1_2025/2025-01-15.csv', 'data_Q1_2025/2025-01-16.csv', 'data_Q1_2025/2025-01-17.csv', 'data_Q1_2025/2025-01-18.csv', 'data_Q1_2025/2025-01-19.csv', 'data_Q1_2025/2025-01-20.csv', 'data_Q1_2025/2025-01-21.csv', 'data_Q1_2025/2025-01-22.csv', 'data_Q1_2025/2025-01-23.csv', 'data_Q1_2025/2025-01-24.csv', 'data_Q1_2025/2025-01-25.csv', 'data_Q1_2025/2025-01-26.csv', 'data_Q1_2025/2025-01-27.csv', 'data_Q1_2025/2025-01-28.csv', 'data_Q1_2025/2025-01-29.csv', 'data_Q1_2025/2025-01-30.csv', 'data_Q1_2025/2025-01-31.csv', 'data_Q1_2025/2025-02-01.csv

In [ ]:
# refactor postgres code to be an executable function
def create_postgres_tbl():
    # Create sql engine
    engine = create_engine('postgresql+psycopg://root:root@localhost:5432/hard_drive_db')

    
    print(pd.io.sql.get_schema(df, name='hard_drive_data', con=engine))

### Complete Data Ingestion Script